### PR TITLE
DFDL Processor Example with Firestore and PubSub

### PR DESCRIPTION
**Pull request by carolinahdez**
Example to process a binary using DFDL definition using Daffodil libraries. The DFDL definition is stored in firestore, the request to process the binary is done through a pubsub sub and the output is published in a JSON format in a Pubsub topic.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,6 +39,7 @@ The examples folder contains example solutions across a variety of Google Cloud
 * [Dataflow Template Pipelines](https://github.com/GoogleCloudPlatform/DataflowTemplates) - Pre-implemented Dataflow template pipelines for solving common data tasks on Google Cloud Platform.
 * [Dataproc GCS Connector](examples/dataproc-gcs-connector) - Install and test unreleased features on the GCS Connector for Dataproc.
 * [Dataproc Persistent History Server for Ephemeral Clusters](examples/dataproc-persistent-history-server) - Example of writing logs from an ephemeral cluster to GCS and using a separate single node cluster to look at Spark and YARN History UIs.
+* [DFDL Processing with Firestore and Pubsub](examples/dfdl-firestore-pubsub-example) - Example to process a binary using DFDL definition using Daffodil libraries. The DFDL definition is stored in firestore, the request to process is done through a pubsub sub and the output is published is a JSON format in Pubsub topic.
```
Comment (iht) [inline README.md:42]: I would include an expansion of the acronym DFDL. Apologies for my ignorance, maybe it is a well known acronym, but I am unfamiliar with it.

File: examples/dfdl-firestore-pubsub-example/README.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
+# DFDL Processor Example
```
Comment (iht) [inline examples/dfdl-firestore-pubsub-example/README.md:1]: What is DFDL? Could the acronym be expanded in this README?

File: examples/dfdl-firestore-pubsub-example/examples/bynary_example.dfdl.xsd [hunk 3]
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (iht) [inline examples/dfdl-firestore-pubsub-example/examples/bynary_example.dfdl.xsd:1]: Copyright header

File: examples/dfdl-firestore-pubsub-example/pom.xml [hunk 4]
```diff
@@ -0,0 +1,107 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (iht) [inline examples/dfdl-firestore-pubsub-example/pom.xml:1]: Copyright header

File: examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlDef.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
+package com.example.dfdl;
```
Comment (iht) [inline examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlDef.java:1]: Copyright header

**Review by iht — CHANGES_REQUESTED** [review]
The copyright header for all the source code files is missing. Please also see minor comment about the acronym used in the example.

**Comment by iht** [comment]
Also, the linter is failing:

```
Step #0 - "lint": Testing formatting for java files in ./examples/dfdl-firestore-pubsub-example
Step #0 - "lint": make: *** [Makefile:28: test] Error 1
Step #0 - "lint": Some files need to be formatted in ./examples/dfdl-firestore-pubsub-example - FAIL
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlService.java
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlDef.java
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/PubSubServer.java
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/ProcessorService.java
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/FirestoreService.java
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/MessageController.java
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlDefRepository.java
```

For instructions about how to run the formatter and the linter in local, please see https://github.com/GoogleCloudPlatform/professional-services/blob/main/CONTRIBUTING.md

**Comment by iht** [comment]
Output of linter:

```
Step #0 - "lint": Some files need to be formatted in ./examples/dfdl-firestore-pubsub-example - FAIL
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlService.java
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlDef.java
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/PubSubServer.java
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/ProcessorService.java
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/FirestoreService.java
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/MessageController.java
Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlDefRepository.java
```

**Review by iht — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (carolinahdez) [reply README.md]: Done

File: examples/dfdl-firestore-pubsub-example/README.md (hunk 2 above)
Comment (carolinahdez) [reply examples/dfdl-firestore-pubsub-example/README.md]: Done

**Comment by carolinahdez** [comment]
> https://github.com/GoogleCloudPlatform/professional-services/blob/main/CONTRIBUTING.md



> Output of linter:
> 
> ```
> Step #0 - "lint": Some files need to be formatted in ./examples/dfdl-firestore-pubsub-example - FAIL
> Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlService.java
> Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlDef.java
> Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/PubSubServer.java
> Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/ProcessorService.java
> Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/FirestoreService.java
> Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/MessageController.java
> Step #0 - "lint": ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlDefRepository.java
> ```

Can you help with this since I am uncertain what I need to fix?

I run the fmt and following is the output:
$ docker run   --mount type=bind,source="$( pwd )",target=/workspace   --workdir=/workspace   gcr.io/$(gcloud config get-value project)/make fmtFormatting ./examples/dfdl-firestore-pubsub-example
Formatting python files (if any)
No python files found for ./examples/dfdl-firestore-pubsub-example - SKIP
Formatting go files (if any)
Formatting java files (if any)
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlDef.java
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/MessageController.java
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/ProcessorService.java
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/FirestoreService.java
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlService.java
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlDefRepository.java
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/PubSubServer.java: could not read file: ./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlDef.java
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/MessageController.java
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/ProcessorService.java
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/FirestoreService.java
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlService.java
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/DfdlDefRepository.java
./examples/dfdl-firestore-pubsub-example/src/main/java/com/example/dfdl/PubSubServer.java

I run the test and it is failing in a directory that has nothing to do with this PR:

l$  docker run   --mount type=bind,source="$( pwd )",target=/workspace   --workdir=/workspace   gcr.io/$(gcloud config get-value project)/make test
Validating ./helpers - Checking bash files

In ./helpers/check_format.sh line 46:
    if [[ ! -z "$FILES_TO_CHECK" ]]
          ^-- SC2236: Use -n instead of ! -z.


In ./helpers/check_format.sh line 56:
	if [[ ! -z "$FILES_TO_LINT"  ]]
              ^-- SC2236: Use -n instead of ! -z.


In ./helpers/check_format.sh line 83:
    if [[ ! -z "$FILES_TO_CHECK" ]]
          ^-- SC2236: Use -n instead of ! -z.


In ./helpers/check_format.sh line 94:
        if [[ ! -z "$FILES_TO_LINT" ]]
              ^-- SC2236: Use -n instead of ! -z.


In ./helpers/check_format.sh line 111:
            if [[ ! -z "$FILES_TO_LINT" ]]
                  ^-- SC2236: Use -n instead of ! -z.


In ./helpers/check_format.sh line 136:
    if [[ ! -z "$FILES_TO_LINT" ]]
          ^-- SC2236: Use -n instead of ! -z.


In ./helpers/check_format.sh line 153:
    if [[ ! -z "$FILES_TO_CHECK" ]]
          ^-- SC2236: Use -n instead of ! -z.


In ./helpers/check_format.sh line 192:
    if [[ ! -z "$FILES_TO_CHECK" ]]
          ^-- SC2236: Use -n instead of ! -z.


In ./helpers/check_format.sh line 203:
        if [[ ! -z "$FILES_TO_LINT" ]]
              ^-- SC2236: Use -n instead of ! -z.

For more information:
  https://www.shellcheck.net/wiki/SC2236 -- Use -n instead of ! -z.

In ./helpers/format.sh line 38:
        if [[ ! -z "$FILES_TO_FORMAT" ]]
              ^-- SC2236: Use -n instead of ! -z.


In ./helpers/format.sh line 72:
        if [[ ! -z "$FILES_TO_FORMAT" ]]
              ^-- SC2236: Use -n instead of ! -z.

For more information:
  https://www.shellcheck.net/wiki/SC2236 -- Use -n instead of ! -z.
Some files need to be formatted in ./helpers - FAIL
./helpers/check_format.sh ./helpers/format.sh 
make: *** [Makefile:28: test] Error 1

**Review by iht — APPROVED** [review]

**Comment by iht** [comment]
This PR was done on top of the old master branch, and it is using a build script that is failing, but the changes pass the build with the current config. Merging PR.

**Comment by iht** [comment]
Thanks for your contribution to the repo!

**Pull request merged**